### PR TITLE
[JENKINS-41102] fix Job "Changes" do not match "Filter for Poll SCM"

### PR DESCRIPTION
**Pull request by ivkrishna-mf**
[Jenkins] RPI 1115588  Job "Changes" do not match "Filter for Poll SCM"

The "Changes" list on my Jenkins jobs (and
corresponding emails) using the AccuRev plugin do not match the change
detection of the job's "Filter for Poll SCM". Promotes to only the
specified sub-folder defined in "Filter for Poll SCM" are correctly
identified and kick-off the job, but the "Changes" list for the job that
gets kicked-off are all changes for the entire stream since the job
last ran, not just the changes for the sub-folder specified in "Filter
for Poll SCM".

My stream contains many sub-folders for each "package", and a Jenkins
job exists for each "package" sub-folder. With the current behavior,
irrelevant change-sets are being sent in job notification emails and
developers not involved with the changes of a specific "package" job are
being notified about success and failures even though they didn't
promote changes to said "package". Is there anyway to make the job
"Changes" list match on the changes based on "Filter for Poll SCM"?
Thanks!

Reference: SI # 3173622 - AccuRev Plugin Fix asked for by Tim Williams
https://issues.jenkins-ci.org/browse/JENKINS-41102

### CODE REVIEW

File: src/main/java/hudson/plugins/accurev/ParseChangeLog.java [hunk 1]
```diff
@@ -84,9 +95,73 @@
       filteredTransactions = transactions;
     }
     filteredTransactions.removeIf(t -> t.getAction().equalsIgnoreCase("dispatch"));
+    
+    logger.info("subpath:" + subpath);
+    if (subpath != null && !subpath.isEmpty()) {
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/ParseChangeLog.java]:
```suggestion
    if (StringUtils.isNotEmpty(subpath)) {
```

File: src/main/java/hudson/plugins/accurev/ParseChangeLog.java [hunk 2]
```diff
@@ -84,9 +95,73 @@
       filteredTransactions = transactions;
     }
     filteredTransactions.removeIf(t -> t.getAction().equalsIgnoreCase("dispatch"));
+    
+    logger.info("subpath:" + subpath);
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/ParseChangeLog.java]: Logging level should be `fine`

File: src/main/java/hudson/plugins/accurev/ParseChangeLog.java [hunk 3]
```diff
@@ -84,9 +95,73 @@
...
+    if (subpath != null && !subpath.isEmpty()) {
+      logger.info("before filter by subpath process..");
+      filteredTransactions.forEach(transaction -> logger.info("transaction:" + transaction));
+      // filtering transaction based on sub-path
+      logger.info("starting filter by subpath process..");
+
+      filteredTransactions = filterBySubpath(filteredTransactions);
+      logger.info("starting filter by subpath process..");
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/ParseChangeLog.java]: This feels like super verbose logging.

File: src/main/java/hudson/plugins/accurev/ParseChangeLog.java [hunk 4]
```diff
@@ -84,9 +95,73 @@
...
+      logger.info("after filter by subpath process..");
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/ParseChangeLog.java]: Shouldn't this be after the `forEach` even though this is super verbose

**Comment by ivkrishna-mf** [comment]
@casz  I have implemented review comments.
can you please check this ?

File: src/main/java/hudson/plugins/accurev/ParseChangeLog.java [hunk 5]
```diff
@@ -84,9 +96,71 @@
...
+
+    List<AccurevTransaction> trans = new ArrayList<>();
+    logger.info("using subpath:" + getSubpath());
+    List<String> subpaths = new ArrayList<String>();
+
+    final StringTokenizer st = new StringTokenizer(getSubpath(), ",");
+    while (st.hasMoreElements()) {
+      String path = st.nextToken().trim();
+      path = path.replace("*", "");
+      logger.info("path:" + path);
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/ParseChangeLog.java]: You're still using log level info here and through out this file 😢

File: src/main/java/hudson/plugins/accurev/delegates/ReftreeDelegate.java [hunk 6]
```diff
@@ -43,6 +44,8 @@ public String getRefTree() {
   protected PollingResult checkForChanges(Job<?, ?> project)
       throws IOException, InterruptedException {
     try {
+      logger.info("setting subpath from Reftree:" + scm.getSubPath());
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/delegates/ReftreeDelegate.java]: and here.

File: src/main/java/hudson/plugins/accurev/delegates/SnapshotDelegate.java [hunk 7]
```diff
@@ -24,6 +25,8 @@ public SnapshotDelegate(AccurevSCM scm) {
 
   private String calculateSnapshotName(final Run<?, ?> build)
       throws IOException, InterruptedException {
+    logger.info("setting subpath from snapshot delegate:" + scm.getSubPath());
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/delegates/SnapshotDelegate.java]: and here.

File: src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java [hunk 8]
```diff
@@ -64,6 +65,8 @@ protected PollingResult checkForChanges(Job<?, ?> project)
     final Date buildDate = lastBuild.getTimestamp().getTime();
     try {
       localStream = scm.getPollingStream(project, listener);
+      logger.info("setting subpath from stream delegate:" + scm.getSubPath());
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java]: and here.

File: src/main/java/hudson/plugins/accurev/delegates/WorkspaceDelegate.java [hunk 9]
```diff
@@ -41,6 +42,8 @@ public String getRefTree() {
   protected PollingResult checkForChanges(Job<?, ?> project)
       throws IOException, InterruptedException, IllegalArgumentException {
     localStream = scm.getPollingStream(project, listener);
+    logger.info("setting subpath from workspace delegate:" + scm.getSubPath());
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/delegates/WorkspaceDelegate.java]: and here.

**Comment by ivkrishna-mf** [comment]
@casz I have changed log level to fine all the places.

**Comment by ivkrishna-mf** [comment]
Thanks. @casz  

I can see oracle jdk used for build accurev plugin. are there any plans for switching to open JDK  since licensing update from Oracle ?

**Comment by casz** [comment]
Jenkins CI is already using AdoptOpenJDK

**Comment by ivkrishna-mf** [comment]
@casz  can you please do a release with this fix today  ? 
And I don't permission to do this.

**Pull request merged**